### ISSUE TITLE
Allow games to have a settings UI before entering a world

### ISSUE THREAD

**Issue by kneekoo** [minetest/minetest#12246]
## Problem

1. Creating a new world with different game settings would be helpful when you want to specify settings unavailable in the Minetest UI.

_In the case of MineClone2, we're [debating how to deal](https://git.minetest.land/MineClone2/MineClone2/issues/2147) with the major mapgen update that Minecraft introduced in 1.18 (deeper and taller worlds), because we will certainly have players who would want to have both "classic" and "new" worlds. Obviously, other settings could be adjusted along with the mapgen type._

2. Entering an existing world could also require a dialog at some point, when the user might have to accept or refuse the conversion of their old worlds into new ones, or when some potentially breaking changes might alter their worlds and they need to be aware of that.

3. Allowing the creation of backups would also be helpful in cases like item 2 mentioned above.

## Solutions

For items 1 and 2, the Minetest UI could offer extra settings that the games provide, but I don't see this as a viable option, considering Minetest can't possibly anticipate what game makers would want their users to tweak before starting a game, which makes it hard or impossible to do a nice UI for everyone. I think it would be much better to leave the UI as it is, and allow the game makers to create their own UIs. Minetest would have to do two things to allow this:
- check if the game has a "pre-init UI"
- if so, don't initiate the world generation/loading until the game says it's OK to fire it up

For item 3, it's probably best not to allow games interact with the filesystem directly, but rather make backups on request - which the games would present as a checkbox for users. This feature would probably need more discussion, but that's out of scope here.

**Comment by Zughy** [comment]
Mmm.. how about having per-world settings that people can tweak from the world selection? That's what I had in mind in my new main menu mockup (world selection, the pencil icon). On the contrary, the gears icon on the left tweak the default game options for every new world

![](https://blog.minetest.net/static/blog/2022_March/menu_concept.png?raw=true)

**Comment by rubenwardy** [comment]
Yeah. My plan for this would be to change Select Mods back into a configure dialog, with the ability for games to customise it. I'd also move the creative and PvP check boxes there

**Comment by kneekoo** [comment]
> Mmm.. how about having per-world settings that people can tweak from the world selection? That's what I had in mind in my new main menu mockup (world selection, the pencil icon). On the contrary, the gears icon on the left tweaks the default game options for every new world

There are two types of settings:
- engine related
- game related

I initially thought that it's best if games design their own UIs, so Minetest doesn't have to worry at all about what game makers want their users to customize before entering a world. Now if the Minetest main menu can acomodate both engine and game related settings before entering a world, that would be really nice.

But I wonder what would happen when games are not updated and the custom settings UI would be either unable to load or it would look broken, or even crash the client. I assume Minetest could reject anything that doesn't conform to the required specifications. Would this work?

**Comment by rubenwardy** [comment]
The description of this issue is a duplicate of #7462

However, the implicit request could be read as "Mainmenu: Add a configure menu that games can customise, replacing Select Mods" which is something new

**Comment by rubenwardy** [comment]
Duplicate of https://github.com/minetest/minetest/issues/5332
